### PR TITLE
Module 1

### PR DESCRIPTION
**Pull request by keliumJU**
**Phoenix LiveView**

_It's a library that abstracts away all the mundane, but important, details so we can focus on adding interactive
real-time features to our app_

-obviously, there has to be some javascript running in the browser

alt text

_Now instead of writing client-side JavaScript, we write server-side Elixir_

**A LiveView is a module with three callback functions**

1.  [mount] assigns the initial state of the LiveViewProcess.
2.  handle event changes the state of the process, and render renders a new view
3.  render for the newly updated states

**Unique things of LiveView**

1. First since LiveViews are server rendered, the original request is just a regular HTTP Request 	so the client gets back a quick initial response of static HTML which has the added benefit of making it SEO friendly
2. Second, since LiveViews use a persistent WebSocket connection after initial request, changes on the server can also be pushed to multiple clientes. that's really inportant for building dustributed real-time applications

### CODE REVIEW

File: live_view_studio/lib/live_view_studio_web/live/light_live.ex [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  """
+  use LiveViewStudioWeb, :live_view
+
+  @doc """
+    #the mount callback it's the first callback that's invoked when the request comes in throught the router
+    *params is a map containing the current query params, as well as any router parameters
+    *session is a map containing private session data
+    *socket is a struct representing the websocket connection
+  """
+  def mount(_params, _session, socket) do
```
Comment (juanhurtado10) [inline live_view_studio/lib/live_view_studio_web/live/light_live.ex:16]: @keliumJU what does a mount do?

File: live_view_studio/lib/live_view_studio_web/live/light_live.ex [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    socket = assign(socket, :brightness, 10)
+    #Calling assign this way returns a new socket struct with :brightness set to 10
+    {:ok, socket}
+  end
+
+  @doc """
+    The render function takes assigns which is the map of key/value pairs
+    we assigned to the socket in mount
+  """
+  def render(assigns) do
```
Comment (juanhurtado10) [inline live_view_studio/lib/live_view_studio_web/live/light_live.ex:27]: @keliumJU what is the purpose of the render callback?

File: live_view_studio/lib/live_view_studio_web/live/light_live.ex [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    """
+  end
+
+  @doc """
+    #The handle_event callback takes three arguments:
+      *the name of the event (on in this case)
+      *metadata related to the event which we’ll ignore
+      *the socket which, remember, has the current state of our live view assigned to it
+  """
+  def handle_event("on", _, socket) do
```
Comment (juanhurtado10) [inline live_view_studio/lib/live_view_studio_web/live/light_live.ex:60]: @keliumJU what does this function do?

File: live_view_studio/lib/live_view_studio_web/router.ex [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    plug :accepts, ["json"]
+  end
+
+  scope "/", LiveViewStudioWeb do
+    pipe_through :browser
+
+    get "/", PageController, :index
+
+    #define live route using the live macro -> takes a path and a LiveView module to use
+    live "/light", LightLive
```
Comment (juanhurtado10) [inline live_view_studio/lib/live_view_studio_web/router.ex:23]: @keliumJU what is the difference between `live` and a regular route as get or post?

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: live_view_studio/lib/live_view_studio_web/live/light_live.ex (hunk 1 above)
Comment (keliumJU) [reply live_view_studio/lib/live_view_studio_web/live/light_live.ex:16]: mount, being the first function that is called in the request, is responsible for opening a connection through websockets, where we have the freedom to create an initial state.

File: live_view_studio/lib/live_view_studio_web/live/light_live.ex (hunk 2 above)
Comment (keliumJU) [reply live_view_studio/lib/live_view_studio_web/live/light_live.ex:27]: render takes care of updating the static files every time the state changes

File: live_view_studio/lib/live_view_studio_web/live/light_live.ex (hunk 3 above)
Comment (keliumJU) [reply live_view_studio/lib/live_view_studio_web/live/light_live.ex:60]: handle_event is responsible for modifying the state each time the "on" function is called, in this case it takes the value of :brightness and updates it to 100

File: live_view_studio/lib/live_view_studio_web/router.ex (hunk 4 above)
Comment (keliumJU) [reply live_view_studio/lib/live_view_studio_web/router.ex:23]: I think the main difference is that live is a webSockets request instead get, post, update, delete ... are HTTP requests

**Review by juanhurtado10 — APPROVED** [review]
Well job done 🚀

**Pull request merged**